File: personal_simple_tcl_minifier/parse.py
def tcl_minify(source: str) -> str:
    tokens: list[str] = _tcl_parse(source)

    _tcl_optimize(tokens)

    return _tcl_unparse(tokens)


def _tcl_parse(source: str) -> list[str]:
    token_start: int = 0
    token_end: int = 0
    source_end: int = len(source)

    tokens: list[str] = []

    while token_start < source_end:
        first_char: str = source[token_start]

        if _is_whitespace_or_semicolon(first_char):
            token_start += 1
            if tokens:
                if _is_whitespace_or_semicolon(tokens[-1]):
                    if first_char == "\n" or (
                        first_char == ";" and tokens[-1] in (" ", "\t")
                    ):
                        tokens[-1] = first_char
                else:
                    tokens.append(first_char)
            continue

        if first_char in ("[", "]", "{", "}", "\\"):
            tokens.append(first_char)
            token_start += 1
            continue

        if _is_token_comment(first_char, tokens):
            token_start = _parse_comment(source, source_end, token_start, tokens)
            continue

        if first_char == '"':
            token_start = _parse_string(source, source_end, token_start, tokens)
            continue

        token_end = token_start + 1
        while token_end < source_end and not _is_delimiter(source[token_end]):
            token_end += 1
        token: str = source[token_start:token_end]
        tokens.append(token)
        token_start = token_end

    return tokens


def _tcl_optimize(tokens: list[str]) -> None:
    if not tokens:
        return

    new_tokens: list[str] = []
    depth: int = 0

    for token in tokens:
        if token == "{":
            depth += 1
        elif token == "}":
            depth -= 1

        if token == "}" and new_tokens and new_tokens[-1].isspace():
            new_tokens[-1] = token
        elif token == "\n" and (not new_tokens or new_tokens[-1] in "\n\\"):
            if new_tokens and new_tokens[-1] == "\\":
                if len(new_tokens) > 1 and new_tokens[-2].isspace():
                    new_tokens.pop()
                else:
                    new_tokens[-1] = " "
        elif _is_token_comment(token, new_tokens):
            # https://wiki.tcl-lang.org/page/Why+can+I+not+place+unmatched+braces+in+Tcl+comments
            if depth > 0:
                depth_change: int = token.count("{") - token.count("}")

                if depth_change:
                    depth += depth_change
                    new_tokens.append(token)
        else:
            new_tokens.append(token)

    while new_tokens and new_tokens[-1].isspace():
        new_tokens.pop()

    tokens[:] = new_tokens


def _tcl_unparse(tokens: list[str]) -> str:
    return "".join(tokens)


def _is_whitespace_or_semicolon(token: str) -> bool:
    return token.isspace() or token == ";"


def _is_delimiter(token: str) -> bool:
    return token.isspace() or token in (";", "[", "]", "{", "}", "\\")


def _is_token_comment(token: str, previous_tokens: list[str]) -> bool:
    return token[0] == "#" and (
        not previous_tokens or previous_tokens[-1][-1:] in ("\n", ";")
    )


def _parse_comment(
    source: str, source_end: int, starting_index: int, out: list[str]
) -> int:
    comment_end: int = source.find("\n", starting_index + 1)
    if comment_end == -1:
        out.append(source[starting_index:])
        return source_end
    else:
        out.append(source[starting_index:comment_end])
        return comment_end


def _parse_string(
    source: str, source_end: int, starting_index: int, out: list[str]
) -> int:
    token_end: int = starting_index + 1
    string_token: str = ""
    local_start: int = starting_index
    skip_whitespace: bool = False

    while token_end < source_end and (
        source[token_end] != '"' or source[token_end - 1] == "\\"
    ):
        if source[token_end].isspace():
            if skip_whitespace:
                token_end += 1
                local_start = token_end
                continue
            if source[token_end - 1] == "\\":
                skip_whitespace = True
                string_token += source[local_start : token_end - 1]
                if not string_token[-1].isspace():
                    string_token += " "
                token_end += 1
                local_start = token_end
                continue

        skip_whitespace = False
        token_end += 1

    token_end += 1
    string_token += source[local_start:token_end]
    out.append(string_token)

    return token_end

File: personal_simple_tcl_minifier/c_extensions/parse.c
#include <ctype.h>
#include <stdbool.h>
#include <stddef.h>
#include <stdlib.h>
#include <string.h>

#include "includes/parse.h"

static inline bool _is_backslash(char c)
{
    return c == '\\';
}

static inline bool _is_whitespace_or_semicolon(char c)
{
    return c == ';' || isspace(c);
}

static inline bool _is_bracket(char c)
{
    switch (c)
    {
    case '{':
    case '}':
    case '[':
    case ']':
        return true;
    default:
        return false;
    }
}

static inline bool _is_delimiter(char c)
{
    switch (c)
    {
    case '\\':
    case '"':
        return true;
    default:
        return _is_bracket(c) || _is_whitespace_or_semicolon(c);
    }
}

static inline bool _is_comment(char c, const char *tcl_minified, int index_minified)
{
    // TODO: See if this can be more generous then just newline/semicolon
    return c == '#' && (index_minified == 0 || tcl_minified[index_minified - 1] == '\n' || tcl_minified[index_minified - 1] == ';');
}

static inline bool _is_string(char c)
{
    return c == '"';
}

#define _append_char_to_minified(c)         \
    {                                       \
        tcl_minified[index_minified++] = c; \
        ++index_source;                     \
    }
#define _append_range_to_minified(s, e)                                      \
    {                                                                        \
        const size_t __token_len = e - s;                                    \
        strncpy(tcl_minified + index_minified, tcl_source + s, __token_len); \
        index_minified += __token_len;                                       \
    }

char *tcl_minify(const char *tcl_source, size_t size, size_t *size_out)
{
    size_t index_source = 0;
    char *tcl_minified = malloc((sizeof(char) * size));
    size_t index_minified = 0;

    int depth = 0;

    while (index_source < size)
    {
        char current_char = tcl_source[index_source];

        if (_is_whitespace_or_semicolon(current_char))
        {
            if (index_minified > 0)
            {
                if (_is_whitespace_or_semicolon(tcl_minified[index_minified - 1]))
                {
                    if (current_char == '\n' || (current_char == ';' && tcl_minified[index_minified - 1] != '\n'))
                    {
                        tcl_minified[index_minified - 1] = current_char;
                    }
                }
                else
                {
                    tcl_minified[index_minified++] = current_char;
                }
            }

            ++index_source;
        }
        else if (_is_bracket(current_char))
        {
            switch (current_char)
            {
            case '{':
                ++depth;
                break;
            case '}':
                if (index_minified > 0 && isspace(tcl_minified[index_minified - 1]))
                {
                    --index_minified;
                }
                --depth;
                break;
            }
            _append_char_to_minified(current_char);
        }
        else if (_is_backslash(current_char))
        {
            ++index_source;
            if (index_source < size)
            {
                if (tcl_source[index_source] == '\n')
                {
                    if (index_minified > 0 && !isspace(tcl_minified[index_minified - 1]))
                    {
                        tcl_minified[index_minified++] = ' ';
                    }

                    while (index_source < size && isspace(tcl_source[index_source]))
                    {
                        ++index_source;
                    }
                }
                else
                {
                    tcl_minified[index_minified++] = '\\';
                    _append_char_to_minified(tcl_source[index_source]);
                }
            }
            else
            {
                tcl_minified[index_minified++] = '\\';
            }
        }
        else if (_is_comment(current_char, tcl_minified, index_minified))
        {
            const size_t start = index_source;
            int open_bracket_count = 0;
            int close_bracket_count = 0;
            while (++index_source < size)
            {
                switch (tcl_source[index_source])
                {
                case '\\':
                    if (index_source + 1 < size)
                    {
                        ++index_source;
                    }
                    break;

                case '\n':
                    goto comment_end_while;

                case '{':
                    ++open_bracket_count;
                    break;
                case '}':
                    ++close_bracket_count;
                    break;
                }
            }

        comment_end_while:
            // https://wiki.tcl-lang.org/page/Why+can+I+not+place+unmatched+braces+in+Tcl+comments
            if (depth > 0 && (open_bracket_count || close_bracket_count))
            {
                depth += open_bracket_count - close_bracket_count;
                _append_range_to_minified(start, index_source)
            }
        }
        else if (_is_string(current_char))
        {
            size_t start = index_source++;
            bool skip_whitespace = false;

            while (index_source < size)
            {
                char string_current_char = tcl_source[index_source];
                if (skip_whitespace && isspace(string_current_char))
                {
                    start = ++index_source;
                    continue;
                }

                switch (string_current_char)
                {
                case '\\':
                    ++index_source;
                    if (index_source < size)
                    {
                        if (tcl_source[index_source] == '\n')
                        {
                            skip_whitespace = true;
                            _append_range_to_minified(start, index_source - 1);
                            if (index_minified > 0 && !isspace(tcl_minified[index_minified - 1]))
                            {
                                _append_char_to_minified(' ');
                            }
                            start = ++index_source;
                        }
                        else
                        {
                            ++index_source;
                        }
                    }
                    break;

                case '"':
                    ++index_source;
                    goto string_while_end;

                default:
                    skip_whitespace = false;
                    ++index_source;
                    break;
                }
            }
        string_while_end:
            _append_range_to_minified(start, index_source)
        }
        else
        {
            const size_t start = index_source++;
            while (index_source < size && !_is_delimiter(tcl_source[index_source]))
            {
                ++index_source;
            }
            _append_range_to_minified(start, index_source)
        }
    }

    while (index_minified > 0 && isspace(tcl_minified[index_minified - 1]))
    {
        --index_minified;
    }

    if (size_out != NULL)
    {
        *size_out = index_minified;
    }
    return tcl_minified;
}